### PR TITLE
Orchestrator zone POST api input validation using annotations

### PR DESCRIPTION
**Pull request by babamishra9**
Orchestrator zone POST api input validation using annotations

- Add validation annotation into OrchestratorZone
  and OrchestratorZonerequest
- Handle as field validation in the OrchestratorZone data class.
- Handle as field validation in OrchestratorZoneRequest data class
- Added exception handlers in controller class
- Controller will return 400 if validation fails.
- Validation error handling message conversion as like ums evolve
- Adjust handle error messages and exceptions in Orchestrator
  Zone controller
- Added ErrorMessageUtil where all message constants, regex patterns
  and utility methods for message conversion are added.
- As request input validation are taken care using annotations in controller and model classes (OrchestratorZoneRequest and
  OrchestratorZone), removed input field validation from service class

### CODE REVIEW

File: uaa/src/test/java/org/cloudfoundry/identity/uaa/integration/OrchestratorZoneControllerIntegrationTests.java [hunk 1]
```diff
@@ -308,27 +312,28 @@ private void testAdminClientSecretAsSpaceAndEmpty(OrchestratorZoneRequest orches
             String.class);
         assertEquals(getResponse.getStatusCode(), HttpStatus.BAD_REQUEST);
         assertEquals(APPLICATION_JSON_UTF8, getResponse.getHeaders().getContentType());
-        String expected  =  JsonUtils.writeValueAsString(
-            new OrchestratorErrorResponse("The adminClientSecret field cannot contain" +
-                                          " spaces or cannot be blank."));
-        assertEquals(expected, getResponse.getBody());
+        assertTrue(getResponse.getBody().contains("orchestratorZoneRequest.parameters.adminClientSecret " +
```
Comment (Yogeshkumarb57) [inline uaa/src/test/java/org/cloudfoundry/identity/uaa/integration/OrchestratorZoneControllerIntegrationTests.java:315]: this one 'orchestratorZoneRequest', i think we decided to remove this as this name is not part of the actual request payload.

File: uaa/src/test/java/org/cloudfoundry/identity/uaa/integration/OrchestratorZoneControllerIntegrationTests.java [hunk 2]
```diff
@@ -308,27 +312,28 @@ private void testAdminClientSecretAsSpaceAndEmpty(OrchestratorZoneRequest orches
...
 
     private void testWithSpaceOrSpecialCharFail(OrchestratorZoneRequest orchestratorZoneRequest) {
         ResponseEntity<String> getResponse = client.exchange(
             serverRunning.getUrl("/orchestrator/zones"), HttpMethod.POST, new HttpEntity<>(orchestratorZoneRequest),
             String.class);
         assertEquals(getResponse.getStatusCode(), HttpStatus.BAD_REQUEST);
-        assertTrue(getResponse.getBody().contains("Special characters are not allowed in the subdomain " +
-                                                  "name except hyphen which can be specified in the middle."));
+        assertEquals(APPLICATION_JSON_UTF8, getResponse.getHeaders().getContentType());
+        assertTrue(getResponse.getBody().contains("orchestratorZoneRequest.parameters.subdomain " +
```
Comment (Yogeshkumarb57) [inline uaa/src/test/java/org/cloudfoundry/identity/uaa/integration/OrchestratorZoneControllerIntegrationTests.java:325]: specifically in this error message 'orchestratorZoneRequest.parameters.subdomain' should not appear at start, it looks weird when we read it.

File: uaa/src/test/java/org/cloudfoundry/identity/uaa/mock/zones/OrchestratorZoneControllerMockMvcTests.java [hunk 3]
```diff
@@ -165,7 +164,7 @@ void testGetZone_nameRequiredError(String url) throws Exception {
     void testGetZone_nameEmptyError(String url) throws Exception {
         performMockMvcCallAndAssertError(get(url), status().isBadRequest(),
                                          JsonUtils.writeValueAsString(
-                                             new OrchestratorErrorResponse("getZone.name: must not be empty")),
+                                             new OrchestratorErrorResponse("getZone.name: must be specified")),
```
Comment (Yogeshkumarb57) [inline uaa/src/test/java/org/cloudfoundry/identity/uaa/mock/zones/OrchestratorZoneControllerMockMvcTests.java:167]: also, here getZone in error message will create confusion between name and getZone.name.

**Review by Yogeshkumarb57 — CHANGES_REQUESTED** [review]

File: zone-service/src/main/java/org/cloudfoundry/identity/uaa/zone/ErrorMessageUtil.java [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+        return objectError instanceof FieldError ? ((FieldError)objectError).getObjectName() + '.' + ((FieldError)objectError).getField() : objectError.getObjectName();
+    }
+
+    public static final String getAffectedProperty(JsonMappingException jsonMappingException) {
+        List path = jsonMappingException.getPath();
+        StringBuilder stringSeparated = new StringBuilder();
+        if(path!= null){
+            for (int i=0; i< path.size(); i++) {
+                String tempStr = (String) path.get(i);
+                stringSeparated.append("\'"+tempStr +"\'");
```
[the remark this answers is not in the document]
Comment (babamishra9) [reply zone-service/src/main/java/org/cloudfoundry/identity/uaa/zone/ErrorMessageUtil.java]: Done. Thanks!

File: zone-service/src/main/java/org/cloudfoundry/identity/uaa/zone/ErrorMessageUtil.java [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+    public static final String ADMIN_CLIENT_CREDENTIALS_VALIDATION_MESSAGE = "must not be empty and must not have empty " +
+                                                                             "spaces";
+    public static final String UAA_CUSTOM_SUBDOMAIN_PATTERN = "(?:[A-Za-z0-9][A-Za-z0-9\\-]{0,61}[A-Za-z0-9]|[A-Za-z0-9])";
+    public static final String UAA_CUSTOM_SUBDOMAIN_MESSAGE = "special characters are not allowed in the" +
+                                                              " subdomain name except hyphen which can be specified in the middle";
+
+    public static final String INVALID_FORMAT_ERROR_MESSAGE = "Request failed due to a validation error";
+
+    public static final String getAffectedProperty(ObjectError objectError) {
+        return objectError instanceof FieldError ? ((FieldError)objectError).getObjectName() + '.' + ((FieldError)objectError).getField() : objectError.getObjectName();
```
[the remark this answers is not in the document]
Comment (babamishra9) [reply zone-service/src/main/java/org/cloudfoundry/identity/uaa/zone/ErrorMessageUtil.java]: Removed ((FieldError)objectError).getObjectName() + '.' +  which will not have orchestratorZoneRequest in the error message. Now the error message response looks like 

{
    "message": "name must not be blank"
}

Done.

File: zone-service/src/main/java/org/cloudfoundry/identity/uaa/zone/OrchestratorZoneController.java [hunk 6]
```diff
@@ -110,4 +109,23 @@ public ResponseEntity<OrchestratorErrorResponse> internalServerError(Exception e
...
+    @ExceptionHandler(value = { MethodArgumentNotValidException.class })
+    public ResponseEntity<OrchestratorErrorResponse> methodArgumentException(MethodArgumentNotValidException ex) {
+        BindingResult bindingResult = ex.getBindingResult();
+        StringBuilder errorMessageBuilder = new StringBuilder();
+        for (ObjectError objectError: bindingResult.getAllErrors()) {
+            errorMessageBuilder = errorMessageBuilder.append(ErrorMessageUtil.getAffectedProperty(objectError)).append(" ");
+            String defaultMessage = objectError.getDefaultMessage();
+            errorMessageBuilder.append(defaultMessage).append(". ; ");
+        }
+        errorMessageBuilder.delete(errorMessageBuilder.length()-4, errorMessageBuilder.length());
```
[the remark this answers is not in the document]
Comment (babamishra9) [reply zone-service/src/main/java/org/cloudfoundry/identity/uaa/zone/OrchestratorZoneController.java]: Done, Thanks!

File: zone-service/src/main/java/org/cloudfoundry/identity/uaa/zone/OrchestratorZoneController.java [hunk 7]
```diff
@@ -110,4 +109,23 @@ public ResponseEntity<OrchestratorErrorResponse> internalServerError(Exception e
     {
         return ResponseEntity.status(INTERNAL_SERVER_ERROR).body(new OrchestratorErrorResponse(ex.getMessage()));
     }
+
+    @ExceptionHandler(value = { HttpMessageNotReadableException.class })
+    public ResponseEntity<OrchestratorErrorResponse> messageReadableException(HttpMessageNotReadableException ex)
+    {
+        return ResponseEntity.status(BAD_REQUEST).body(new OrchestratorErrorResponse(ErrorMessageUtil.getErrorDetail(ex)));
```
[the remark this answers is not in the document]
Comment (babamishra9) [reply zone-service/src/main/java/org/cloudfoundry/identity/uaa/zone/OrchestratorZoneController.java:119]: Done. Added below details to commit message. Thanks!

Added below exception handlers for input request validation.
    - messageReadableException : this is for miss match data,
      if we are passing any wrong data like for example
       "[[[[]]]]==---222"
      then HttpMessageNotReadableException exception will be thrown
      by spring validator while converting the request body.
    - methodArgumentException : this for the data validation along with
      patterns defined in property annotation.
      MethodArgumentNotValidException exception will be thrown.
    - constraintViolationException : this is when any request parameter
      data are validated by framework which by default add the
      methodname.property in the error message for example
      "getZone.name: must be specified", so explicitly methodname
      prefix has been removed to customize the error message.

File: uaa/src/test/java/org/cloudfoundry/identity/uaa/integration/OrchestratorZoneControllerIntegrationTests.java (hunk 1 above)
Comment (babamishra9) [reply uaa/src/test/java/org/cloudfoundry/identity/uaa/integration/OrchestratorZoneControllerIntegrationTests.java]: Done. Thanks!

File: uaa/src/test/java/org/cloudfoundry/identity/uaa/integration/OrchestratorZoneControllerIntegrationTests.java (hunk 2 above)
Comment (babamishra9) [reply uaa/src/test/java/org/cloudfoundry/identity/uaa/integration/OrchestratorZoneControllerIntegrationTests.java]: Done. Thanks!

File: uaa/src/test/java/org/cloudfoundry/identity/uaa/mock/zones/OrchestratorZoneControllerMockMvcTests.java (hunk 3 above)
Comment (babamishra9) [reply uaa/src/test/java/org/cloudfoundry/identity/uaa/mock/zones/OrchestratorZoneControllerMockMvcTests.java]: Done. Thanks!

File: zone-service/src/main/java/org/cloudfoundry/identity/uaa/zone/OrchestratorZoneController.java [hunk 8]
```diff
@@ -110,4 +112,29 @@ public ResponseEntity<OrchestratorErrorResponse> internalServerError(Exception e
...
+            errorMessages.add(message);
+        }
+        return ResponseEntity.status(BAD_REQUEST).body(new OrchestratorErrorResponse(String.join("; ",
+                                                                                                 errorMessages)));
+    }
+
+    @ExceptionHandler(value = { ConstraintViolationException.class})
+    public ResponseEntity<OrchestratorErrorResponse> constraintViolationException(ConstraintViolationException ex)
+    {
+        String errorMessage = ex.getMessage().replaceAll(GET_ZONE_PREFIX, "");
```
Comment (Yogeshkumarb57) [inline zone-service/src/main/java/org/cloudfoundry/identity/uaa/zone/OrchestratorZoneController.java:137]: May be handle  ConstraintViolationException in this way https://stackoverflow.com/questions/36555057/get-field-name-when-javax-validation-constraintviolationexception-is-thrown and https://www.baeldung.com/global-error-handler-in-a-spring-rest-api.

File: uaa/src/test/java/org/cloudfoundry/identity/uaa/mock/zones/OrchestratorZoneControllerMockMvcTests.java [hunk 9]
```diff
@@ -165,7 +164,7 @@ void testGetZone_nameRequiredError(String url) throws Exception {
     void testGetZone_nameEmptyError(String url) throws Exception {
         performMockMvcCallAndAssertError(get(url), status().isBadRequest(),
                                          JsonUtils.writeValueAsString(
-                                             new OrchestratorErrorResponse("getZone.name: must not be empty")),
+                                             new OrchestratorErrorResponse("name: must be specified")),
```
Comment (Yogeshkumarb57) [inline uaa/src/test/java/org/cloudfoundry/identity/uaa/mock/zones/OrchestratorZoneControllerMockMvcTests.java:167]: Not sure about this error message "name: must be specified", how this is coming? from api response or just test data but all error messages should follow same format. Include or remove ":" from all error messages.

File: zone-service/src/main/java/org/cloudfoundry/identity/uaa/zone/OrchestratorZoneController.java (hunk 8 above)
Comment (babamishra9) [reply zone-service/src/main/java/org/cloudfoundry/identity/uaa/zone/OrchestratorZoneController.java]: Done, Thanks!
which will retrun error response as {
    "message": "name must be specified"
} for both get and delete zone.
Also added a integration test for testDeleteZone_NameRequiredError()

File: uaa/src/test/java/org/cloudfoundry/identity/uaa/mock/zones/OrchestratorZoneControllerMockMvcTests.java (hunk 9 above)
Comment (babamishra9) [reply uaa/src/test/java/org/cloudfoundry/identity/uaa/mock/zones/OrchestratorZoneControllerMockMvcTests.java:167]: Done . Thanks!

Retrun error response as {
"message": "name must be specified"
} for both get and delete zone.

File: uaa/src/test/java/org/cloudfoundry/identity/uaa/integration/OrchestratorZoneControllerIntegrationTests.java [hunk 10]
```diff
@@ -308,27 +330,28 @@ private void testAdminClientSecretAsSpaceAndEmpty(OrchestratorZoneRequest orches
...
         ResponseEntity<String> getResponse = client.exchange(
             serverRunning.getUrl("/orchestrator/zones"), HttpMethod.POST, new HttpEntity<>(orchestratorZoneRequest),
             String.class);
         assertEquals(getResponse.getStatusCode(), HttpStatus.BAD_REQUEST);
-        assertTrue(getResponse.getBody().contains("Special characters are not allowed in the subdomain " +
-                                                  "name except hyphen which can be specified in the middle."));
+        assertEquals(APPLICATION_JSON_UTF8, getResponse.getHeaders().getContentType());
+        assertTrue(getResponse.getBody().contains("parameters.subdomain " +
+                                                  "special characters are not allowed in the subdomain name " +
+                                                  "except hyphen which can be specified in the middle"));
```
Comment (Yogeshkumarb57) [inline uaa/src/test/java/org/cloudfoundry/identity/uaa/integration/OrchestratorZoneControllerIntegrationTests.java:345]: @babamishra9 Please confirm this error message with pushkaraj for its format.

Comment (babamishra9) [reply uaa/src/test/java/org/cloudfoundry/identity/uaa/integration/OrchestratorZoneControllerIntegrationTests.java]: As discussed, error messages are updated.

File: uaa/src/test/java/org/cloudfoundry/identity/uaa/mock/zones/OrchestratorZoneControllerMockMvcTests.java [hunk 11]
```diff
@@ -375,6 +373,91 @@ void testCreateZone_Accepted_Success() throws Exception {
...
+                             "  \"parameters\": {\n" +
+                             "    \"adminClientSecret\": \"dsfds\",\n" +
+                             "    \"subdomain\" : \"test-zone-0\"\n" +
+                             "  }\n" +
+                             "}")
+                    .header("Authorization", "Bearer " + uaaAdminClientToken))
+            .andExpect(status().isBadRequest()).andReturn();
+
+        assertEquals(APPLICATION_JSON_VALUE, result.getResponse().getContentType());
+        assertTrue(result.getResponse().getContentAsString().contains("name Cannot deserialize value of type " +
```
Comment (Yogeshkumarb57) [inline uaa/src/test/java/org/cloudfoundry/identity/uaa/mock/zones/OrchestratorZoneControllerMockMvcTests.java:427]: Same casing thing here with `Cannot` word.

Comment (babamishra9) [reply uaa/src/test/java/org/cloudfoundry/identity/uaa/mock/zones/OrchestratorZoneControllerMockMvcTests.java]: Done.

File: uaa/src/test/java/org/cloudfoundry/identity/uaa/mock/zones/OrchestratorZoneControllerMockMvcTests.java [hunk 12]
```diff
@@ -375,6 +373,91 @@ void testCreateZone_Accepted_Success() throws Exception {
...
+                             "  \"parameters\": {\n" +
+                             "    \"adminClientSecret\": 0992932.303203.00223\n" +
+                             "    \"subdomain\" : \"uywyyw\"\n" +
+                             "  }\n" +
+                             "}")
+                    .header("Authorization", "Bearer " + uaaAdminClientToken))
+            .andExpect(status().isBadRequest()).andReturn();
+
+        assertEquals(APPLICATION_JSON_VALUE, result.getResponse().getContentType());
+        assertTrue(result.getResponse().getContentAsString().contains("parameters Invalid numeric value"));
```
[the remark this answers is not in the document]
Comment (babamishra9) [reply uaa/src/test/java/org/cloudfoundry/identity/uaa/mock/zones/OrchestratorZoneControllerMockMvcTests.java]: "I" as capital letter comes as default from the exception original message, as a work around, first letter is now converted to lowercase with manual adjustment using Character.toLowerCase(errorMessage.charAt(0)) + errorMessage.substring(1).

As discussed, parameters is the only field name which is showing in error.getFieldName(), I think while parsing the json to model type mismatch results null for parameters as we were debugging the value for error-> path and description.  I think its ok to have the same message. If you have anything please suggest.

[the remark this answers is not in the document]
Comment (babamishra9) [reply uaa/src/test/java/org/cloudfoundry/identity/uaa/mock/zones/OrchestratorZoneControllerMockMvcTests.java]: As discussed over daily scrum call, removed "Character.toLowerCase(errorMessage.charAt(0)) + errorMessage.substring(1) from the code and added "  is invalid: " string[https://github.com/GESoftware-CF/uaa/pull/517/commits/97bd4b5fa0bfcea41c5b0c7fd72661147d79604c#diff-bb1faf425c2559b692ad6cecdb58954effa340801ace6ac0bc460e2329bc73dcR75](url) in the now the error message check in testcase have the error message assertions as "parameters is invalid: Invalid numeric value".

File: uaa/src/test/java/org/cloudfoundry/identity/uaa/mock/zones/OrchestratorZoneControllerMockMvcTests.java (hunk 11 above)
Comment (babamishra9) [reply uaa/src/test/java/org/cloudfoundry/identity/uaa/mock/zones/OrchestratorZoneControllerMockMvcTests.java]: As discussed over daily scrum call, removed "Character.toLowerCase(errorMessage.charAt(0)) + errorMessage.substring(1) from the code and added " is invalid: " string

**Review by Yogeshkumarb57 — COMMENTED** [review]
@babamishra9 I'm good with this PR changes now but please fix conflicting files.

**Comment by babamishra9** [comment]
> @babamishra9 I'm good with this PR changes now but please fix conflicting files.

@Yogeshkumarb57 , I have merged predix_extensions_75.18.1 to current feature branch validation_using_annotation. Please review and approve. Thanks!

**Review by Yogeshkumarb57 — APPROVED** [review]
@babamishra9 Approving PR, please check build status before merging.

**Pull request merged**